File: SEARCH/Binary_search.cpp
#include <iostream>
#include <algorithm>
#include <cmath>

// Определение разера массива по указателям
template <typename Type>
size_t len(Type *begin, Type *end) {
	Type *cur = begin;
	size_t count = 0;
	if (begin == end) { return 1; }
	while (cur != end) {
		count++;
		cur++;
	}
	return count;
}

 /***************Бинарный поиск - binary_search***************/
template <typename Type>
size_t binary_search(Type *first, Type *last, const Type &elm) {
	
	size_t mid;
	size_t id_beg = 0;
	size_t id_end = len(first, last);

	while (id_beg <= id_end) {
		mid = id_beg + (id_end - id_beg) / 2;
		if (elm == first[mid]) {
			return mid;
		}
		else if (elm > first[mid]) {
			id_beg = mid + 1;
		}
		else {
			id_end = mid - 1;
		}
	}
	return 0;
}
/**************************************************************/
// Рекурсивынй вариант
template <typename Type>
size_t binary_search_rec(Type *arr, const Type &value, size_t begin = 0, size_t end = 0) {

	if (begin > end) { return -1; }
	size_t mid = begin + (end - begin) / 2;	// Формула, чтобы не было переполнения
	if (value == arr[mid]) {
		return mid;
	}
	if (value < arr[mid]) {
		return binary_search_rec(arr, value, begin, mid - 1);
	}
	return binary_search_rec(arr, value, mid + 1, end);
}

// Для контейнеров с логикой итераторов
template<class ForwardIt, class T>
bool spbinary_search(ForwardIt first, ForwardIt last, const T& value) {
	first = std::lower_bound(first, last, value);
	return (!(first == last) && !(value < *first));
}

// Генерация случайных чаров для теста
char random_char(int range) {
	const int MAX = 26;
	char alpha[MAX] = { 'a', 'b', 'c', 'd', 'e', 'f', 'g',
						  'h', 'i', 'j', 'k', 'l', 'm', 'n',
						  'o', 'p', 'q', 'r', 's', 't', 'u',
						  'v', 'w', 'x', 'y', 'z' };
	std::string result = "";
	for (int i = 0; i < range; i++)
		result = result + alpha[rand() % MAX];

	return result[range % MAX];
}

// Тесты времени пойска в отсортированных контейнерах
void time_test(size_t length) {
	setlocale(LC_ALL, "Ru");
	size_t  size = length;
	size_t *uint_array = new size_t[size];
	double *doub_array = new double[size];
	char   *char_array = new char[size];
	char	ch = 'a';

	for (size_t i = 0; i < size; ++i) {
		uint_array[i] = rand() % RAND_MAX;
		doub_array[i] = round(rand() % RAND_MAX * 0.003) / 100;
		char_array[i] = char(ch);
		if (ch != 'z') { ch++; }
		else ch = 'a';
	}
	// Сортировка массивов
	std::sort(uint_array, uint_array + size);
	std::sort(doub_array, doub_array + size);
	std::sort(char_array, char_array + size);
	// Тесты итеративных версий
	clock_t timer = clock();
	size_t uarr = binary_search(uint_array, uint_array + size, static_cast<size_t>(rand() % RAND_MAX));
	clock_t ui_time = clock() - timer;
	
	timer = clock();
	size_t darr = binary_search(doub_array, doub_array + size, round(rand() % RAND_MAX * 0.003) / 100);
	clock_t do_time = clock() - timer;
	
	timer = clock();
	size_t carr = binary_search(char_array, char_array + size, random_char(50));
	clock_t ch_time = clock() - timer;

	std::cout << "1. Integer_id = " << uarr << "; Time_search: " << double(ui_time) / CLOCKS_PER_SEC << " sec.\n";
	std::cout << "2. Float_id = "   << darr << "; Time_search: " << double(do_time) / CLOCKS_PER_SEC << " sec.\n";
	std::cout << "3. Char_id = "    << carr << "; Time_search: " << double(ch_time) / CLOCKS_PER_SEC << " sec.\n";
}


int main() {

	time_test(1'000'000);
}

File: SEARCH/Binary_search.py
import random
import time

"""**********Итеративный бинарный поиск - binary_search**********"""
def binary_search(array, item):
    low = 0                                 # Начало поиска (указатель на 0-ой индекс массива)
    high = len(array) - 1                   # Конец поиска (указатель на последний элемент)

    while low <= high:                      # Пока указатели не встретились, выполнять следующую логику
        mid = low + (high - low) // 2       # Индекс середины массива
        if item == array[mid]:              # Если искомый элемент найден, вернуть индекс этого элемента
            return mid
        elif item > array[mid]:             # Если искомый больше среднего, значит искать справа
            low = mid + 1                   # Левый указатель теперь указывает на середину массива - 1
        else:                               # Если искомый элемент меньше среднего
            high = mid - 1                  # Сдвинуть правый указатель на середину - 1
    return None                             # Если элемент не нашёлся - вернуть ничего
"""**************************************************************"""

"""**********Рекурсивный бинарный поиск - binary_search**********"""
def binary_search_recursive(array, elem, start=0, end=None):
    if end is None:
        end = len(array) - 1
    if start > end:
        return None
    mid = (start + end) // 2
    if elem == array[mid]:
        return mid
    if elem < array[mid]:
        return binary_search_recursive(array, elem, start, mid - 1)
    # elem > arr[mid]
    return binary_search_recursive(array, elem, mid + 1, end)
"""**************************************************************"""

# Функция тестирования
def time_test(length):
    array_int = [round(random.random()*100) for i in range(length)]
    array_float = [round(random.random()*100, 2) for i in range(length)]
    array_char = [chr(random.randint(ord('a'), ord('z'))) for i in range(length)]

    array_int.sort()
    array_float.sort()
    array_char.sort()

    timer = time.perf_counter()
    id_int = binary_search(array_int, round(random.random()*100))
    t_int = time.perf_counter() - timer

    timer = time.perf_counter()
    id_float = binary_search(array_float, round(random.random()*100, 2))
    t_float = time.perf_counter() - timer

    tr = time.perf_counter()
    id_char = binary_search(array_char, chr(random.randint(ord('a'), ord('z'))))
    t_char = time.perf_counter() - tr

    print(f"Integer_id = {id_int} --> time: {t_int:0.5f}\nFloat_id = {id_float} --> time: {t_float:0.5f}")
    print(f"Char_id = {id_char} --> time: {t_char:0.5f}")


if __name__ == "__main__":
    #main()
    time_test(100_000)